### PR TITLE
Adds filter as request for  #33

### LINKED ISSUE #33: win_disk_facts: Add a `subset` parameter to gather only specific facts
_From @chopraaa on Feb 05, 2019 21:31_

##### ISSUE TYPE
- Feature Idea

##### COMPONENT NAME
win_disk_facts

##### ADDITIONAL INFORMATION
This can be used to reduce the data gathered and returned by the `win_disk_facts` module.

For a single disk/single partition server, this returns:

```json
    "ansible_facts": {
        "ansible_disks": [
            {
                "bootable": true,
                "bus_type": "SAS",
                "clustered": false,
                "firmware_version": "2.0 ",
                "friendly_name": "AWS PVDISK SCSI Disk Device",
                "guid": null,
                "location": "PCIROOT(0)#PCI(0300)#PCI(0300)#SAS(P00T00L00)",
                "manufacturer": "AWS     ",
                "model": "PVDISK          ",
                "number": 0,
                "operational_status": "Online",
                "partition_count": 2,
                "partition_style": "MBR",
                "partitions": [
                    {
                        "access_paths": [
                            "\\\\?\\Volume{74201446-2987-11e9-825e-806e6f6e6963}\\"
                        ],
                        "active": true,
                        "drive_letter": null,
                        "guid": null,
                        "hidden": false,
                        "mbr_type": 7,
                        "number": 1,
                        "offset": 1048576,
                        "shadow_copy": false,
                        "size": 367001600,
                        "transition_state": 1,
                        "type": "IFS",
                        "volumes": [
                            {
                                "allocation_unit_size": 4096,
                                "drive_type": "Fixed",
                                "health_status": "Healthy",
                                "label": "",
                                "object_id": "\\\\?\\Volume{74201446-2987-11e9-825e-806e6f6e6963}\\",
                                "path": "\\\\?\\Volume{74201446-2987-11e9-825e-806e6f6e6963}\\",
                                "size": 366997504,
                                "size_remaining": 92209152,
                                "type": "NTFS"
                            }
                        ]
                    },
                    {
                        "access_paths": [
                            "C:\\",
                            "\\\\?\\Volume{74201447-2987-11e9-825e-806e6f6e6963}\\"
                        ],
                        "active": false,
                        "drive_letter": "C",
                        "guid": null,
                        "hidden": false,
                        "mbr_type": 7,
                        "number": 2,
                        "offset": 368050176,
                        "shadow_copy": false,
                        "size": 31843155968,
                        "transition_state": 1,
                        "type": "IFS",
                        "volumes": [
                            {
                                "allocation_unit_size": 4096,
                                "drive_type": "Fixed",
                                "health_status": "Healthy",
                                "label": "",
                                "object_id": "\\\\?\\Volume{74201447-2987-11e9-825e-806e6f6e6963}\\",
                                "path": "\\\\?\\Volume{74201447-2987-11e9-825e-806e6f6e6963}\\",
                                "size": 31843151872,
                                "size_remaining": 9631539200,
                                "type": "NTFS"
                            }
                        ]
                    }
                ],
                "path": "\\\\?\\scsi#disk&ven_aws&prod_pvdisk#000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}",
                "physical_disk": {
                    "allocated_size": 32212254720,
                    "bus_type": "SAS",
                    "can_pool": false,
                    "cannot_pool_reason": "Insufficient Capacity",
                    "device_id": "0",
                    "firmware_version": "2.0 ",
                    "friendly_name": "PhysicalDisk0",
                    "health_status": "Healthy",
                    "indication_enabled": null,
                    "manufacturer": "AWS     ",
                    "media_type": "UnSpecified",
                    "model": "PVDISK          ",
                    "object_id": "{1}\\\\WIN-PE4KPP7H947\\root/Microsoft/Windows/Storage/Providers_v2\\SPACES_PhysicalDisk.ObjectId=\"{74201442-2987-11e9-825e-806e6f6e6963}:PD:{09573914-0d84-efea-b97b-68ecb801c634}\"",
                    "operational_status": "OK",
                    "partial": true,
                    "physical_location": null,
                    "serial_number": "vol02f99e3a82c0653a4",
                    "size": 32212254720,
                    "spindle_speed": 4294967295,
                    "supported_usages": {
                        "Count": 5,
                        "value": [
                            "Auto-Select",
                            "Manual-Select",
                            "Hot Spare",
                            "Retired",
                            "Journal"
                        ]
                    },
                    "unique_id": "vol02f99e3a82c0653a4",
                    "usage_type": "Auto-Select"
                },
                "read_only": false,
                "sector_size": 512,
                "serial_number": "vol02f99e3a82c0653a4",
                "size": 32212254720,
                "system_disk": true,
                "unique_id": "vol02f99e3a82c0653a4"
            }
        ]
    },
    "changed": false
}
```

And for servers with multiple disks and volumes (I deal with 20+ volumes on a single server) this can get pretty ugly and slow to process. In addition to this, you need to use `set_fact` once and iterate through that again to get specific values.

I suggest adding a `filter` parameter to output specific data. Usage for listing `volumes` and `dynamic_disks`:

```yaml
win_disk_facts:
  filter:
    - volumes
    - dynamic_disks
```

_Copied from original issue: ansible/ansible#51770_

### PR DESCRIPTION
**Pull request by gamethis**
##### SUMMARY

This allows for selecting 1 or more filter option from the following choices
``` physical_disk, virtual_disk, win32_disk_drive, partitions, volumes ```

Filter is option and by default selects all choices which gives same behavior as original module.

fixes #33 

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
win_disk_facts

##### ADDITIONAL INFORMATION
Adds filter option to win_disk_facts

```
- name: get disk physical_disk facts on the target
  community.windows.win_disk_facts:
    filter:
      - partitions

```
before change output
```
{"ansible_facts": {"ansible_disks": [{"bootable": true, "bus_type": "SAS", "clustered": false, "firmware_version": "1.0 ", "friendly_name": "VMware, VMware Virtual S", "guid": null, "location": "SCSI0", "manufacturer": "VMware, ", "model": "VMware Virtual S", "number": 0, "operational_status": "Online", "partition_count": 1, "partition_style": "MBR", "partitions": [{"access_paths": ["C:\\", "\\\\?\\Volume{7aa64acc-0000-0000-0000-100000000000}\\"], "active": true, "drive_letter": "C", "guid": null, "hidden": false, "mbr_type": 7, "number": 1, "offset": 1048576, "shadow_copy": false, "size": 107372085248, "transition_state": 1, "type": "IFS", "volumes": [{"allocation_unit_size": 4096, "drive_type": "Fixed", "health_status": "Healthy", "label": "OS", "object_id": "{1}\\\\THDTHE-7HGCMSJE\\root/Microsoft/Windows/Storage/Providers_v2\\WSP_Volume.ObjectId=\"{944bfc4d-16a7-11ec-aeb2-806e6f6e6963}:VO:\\\\?\\Volume{7aa64acc-0000-0000-0000-100000000000}\\\"", "path": "\\\\?\\Volume{7aa64acc-0000-0000-0000-100000000000}\\", "size": 107372081152, "size_remaining": 94551818240, "type": "NTFS"}]}], "path": "\\\\?\\scsi#disk&ven_vmware_&prod_vmware_virtual_s#000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}", "physical_disk": {"allocated_size": 107374182400, "bus_type": "SAS", "can_pool": false, "cannot_pool_reason": "Insufficient Capacity", "device_id": "0", "firmware_version": "1.0", "friendly_name": "VMware, VMware Virtual S", "health_status": "Healthy", "indication_enabled": null, "manufacturer": "VMware,", "media_type": "SSD", "model": "VMware Virtual S", "object_id": "{1}\\\\THDTHE-7HGCMSJE\\root/Microsoft/Windows/Storage/Providers_v2\\SPACES_PhysicalDisk.ObjectId=\"{944bfc4d-16a7-11ec-aeb2-806e6f6e6963}:PD:{25d8018f-59b6-c8c8-a34a-315de3c3865e}\"", "operational_status": "OK", "partial": true, "physical_location": "SCSI0", "serial_number": "6000c29ebffbb2bb9b1a16a3643a7932", "size": 107374182400, "spindle_speed": 0, "supported_usages": {"Count": 5, "value": ["Auto-Select", "Manual-Select", "Hot Spare", "Retired", "Journal"]}, "unique_id": "6000C29EBFFBB2BB9B1A16A3643A7932", "usage_type": "Auto-Select"}, "read_only": false, "sector_size": 512, "serial_number": "6000c29ebffbb2bb9b1a16a3643a7932", "size": 107374182400, "system_disk": true, "unique_id": "6000C29EBFFBB2BB9B1A16A3643A7932", "win32_disk_drive": {"availability": null, "bytes_per_sector": 512, "capabilities": [3, 4], "capability_descriptions": ["Random Access", "Supports Writing"], "caption": "VMware, VMware Virtual S SCSI Disk Device", "compression_method": null, "config_manager_error_code": 0, "config_manager_user_config": false, "creation_class_name": "Win32_DiskDrive", "default_block_size": null, "description": "Disk drive", "device_id": "\\\\.\\PHYSICALDRIVE0", "error_cleared": null, "error_description": null, "error_methodology": null, "firmware_revision": "1.0 ", "index": 0, "install_date": null, "interface_type": "SCSI", "last_error_code": null, "manufacturer": "(Standard disk drives)", "max_block_size": null, "max_media_size": null, "media_loaded": true, "media_type": "Fixed hard disk media", "min_block_size": null, "model": "VMware, VMware Virtual S SCSI Disk Device", "name": "\\\\.\\PHYSICALDRIVE0", "needs_cleaning": null, "number_of_media_supported": null, "partitions": 1, "pnp_device_id": "SCSI\\DISK&VEN_VMWARE_&PROD_VMWARE_VIRTUAL_S\\000000", "power_management_capabilities": null, "power_management_supported": null, "scsi_bus": 0, "scsi_logical_unit": 0, "scsi_port": 2, "scsi_target_id": 0, "sectors_per_track": 63, "serial_number": "6000c29ebffbb2bb9b1a16a3643a7932", "signature": 2057718476, "size": 107372805120, "status": "OK", "status_info": null, "system_creation_class_name": "Win32_ComputerSystem", "system_name": "THDTHE-7HGCMSJE", "total_cylinders": 13054, "total_heads": 255, "total_sectors": 209712510, "total_tracks": 3328770, "tracks_per_cylinder": 255}}]}, "changed": false}
```

after change output
```
{"ansible_facts": {"ansible_disks": [{"bootable": true, "bus_type": "SAS", "clustered": false, "firmware_version": "1.0 ", "friendly_name": "VMware, VMware Virtual S", "guid": null, "location": "SCSI0", "manufacturer": "VMware, ", "model": "VMware Virtual S", "number": 0, "operational_status": "Online", "partition_count": 1, "partition_style": "MBR", "partitions": [{"access_paths": ["C:\\", "\\\\?\\Volume{7aa64acc-0000-0000-0000-100000000000}\\"], "active": true, "drive_letter": "C", "guid": null, "hidden": false, "mbr_type": 7, "number": 1, "offset": 1048576, "shadow_copy": false, "size": 107372085248, "transition_state": 1, "type": "IFS"}], "path": "\\\\?\\scsi#disk&ven_vmware_&prod_vmware_virtual_s#000000#{53f56307-b6bf-11d0-94f2-00a0c91efb8b}", "read_only": false, "sector_size": 512, "serial_number": "6000c29ebffbb2bb9b1a16a3643a7932", "size": 107374182400, "system_disk": true, "unique_id": "6000C29EBFFBB2BB9B1A16A3643A7932"}]}, "changed": false}
```

### CODE REVIEW

File: plugins/modules/win_disk_facts.ps1 [hunk 1]
```diff
@@ -4,16 +4,29 @@
 # GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
 
 #Requires -Module Ansible.ModuleUtils.Legacy
+#AnsibleRequires -CSharpUtil Ansible.Basic
```
Comment (jborean93) [inline plugins/modules/win_disk_facts.ps1:7]: If you are converting this to the newer style the above line `#Requires -Module Ansible.ModuleUtils.Legacy` is no longer needed but it requires a few more changes to migrate from the legacy stuff. The results are meant to be stored in `$module.Result`, `Exit-Json` and `Fail-Json` have changed to `$module.ExitJson()` and `$module.FailJson("msg")`. Unless you are willing to do all this work I would stick with the legacy style for now or address the migration to `Ansible.Basic` in a separate PR.

**Review by jborean93 — CHANGES_REQUESTED** [review]

File: plugins/modules/win_disk_facts.py [hunk 2]
```diff
@@ -18,6 +18,15 @@
...
   - Marc Tschapek (@marqelme)
+options:
+  filter:
+    description:
+      - Allows to filter returned facts by type of disk information.
+      - If volumes are selected partitions will be returned as well.
+    type: list
+    elements: str
+    choices: [ physical_disk, virtual_disk, win32_disk_drive, partitions, volumes ]
+    default: [ physical_disk, virtual_disk, win32_disk_drive, partitions, volumes ]
```
Comment (jborean93) [inline plugins/modules/win_disk_facts.py:29]: Needs `version_added: 1.9.0` to denote when this option was introduced in the collection.

**Review by jborean93 — COMMENTED** [review]
Changes look good, if you get a chance it would be great if you could add a changelog fragment, I've shared some details on this in another one of your PRs https://github.com/ansible-collections/community.windows/pull/337#pullrequestreview-811969811.

File: plugins/modules/win_disk_facts.py (hunk 2 above)
Comment (gamethis) [reply plugins/modules/win_disk_facts.py:29]: updated doc and added change fragment.

**Review by jborean93 — APPROVED** [review]

**Pull request merged**